107/src/main/java/org/ehcache/jsr107/EhcacheCachingProvider.java

```diff
@@ -156,6 +156,61 @@ private Eh107CacheManager createCacheManager(URI uri, Configuration config, Prop
             new ConfigurationMerger(config, jsr107Service, cacheLoaderWriterFactory));
   }
 
+  /**
+   * Enables to create a JSR-107 {@link CacheManager} based on the provided Ehcache {@link CacheManagerBuilder}.
+   *
+   * @param uri the URI identifying this cache manager
+   * @param classLoader class loader where the cache manager will be registered
+   * @param builder the builder to create the underlying EhcacheManager
+   *
+   * @return a cache manager
+   */
+  public CacheManager getCacheManager(URI uri, ClassLoader classLoader, CacheManagerBuilder<org.ehcache.CacheManager> builder) {
+    Eh107CacheManager cacheManager;
+
+    synchronized (cacheManagers) {
+      ConcurrentMap<URI, Eh107CacheManager> byURI = cacheManagers.get(classLoader);
+      if (byURI == null) {
+        byURI = new ConcurrentHashMap<>();
+        cacheManagers.put(classLoader, byURI);
+      }
+
+      cacheManager = byURI.get(uri);
+      if (cacheManager == null || cacheManager.isClosed()) {
+
+        if(cacheManager != null) {
+          byURI.remove(uri, cacheManager);
+        }
+
+        cacheManager = createCacheManager(uri, classLoader, builder);
+        byURI.put(uri, cacheManager);
+      }
+    }
+
+    return cacheManager;
+  }
+
+  private Eh107CacheManager createCacheManager(URI uri, ClassLoader classLoader, CacheManagerBuilder<org.ehcache.CacheManager> builder) {
+    Eh107CacheLoaderWriterProvider cacheLoaderWriterFactory = new Eh107CacheLoaderWriterProvider();
+
+    Object[] serviceCreationConfigurations = builder.getServices().toArray();
+
+    Jsr107Service jsr107Service = new DefaultJsr107Service(ServiceUtils.findSingletonAmongst(Jsr107Configuration.class, serviceCreationConfigurations));
+
+    Collection<Service> services = new ArrayList<>(4);
+    services.add(cacheLoaderWriterFactory);
+    services.add(jsr107Service);
+
+    if (ServiceUtils.findSingletonAmongst(DefaultSerializationProviderConfiguration.class, serviceCreationConfigurations) == null) {
+      services.add(new DefaultJsr107SerializationProvider());
+    }
+
+    InternalCacheManager ehcacheManager = (InternalCacheManager) builder.build(true);
```

Comment:
Need to actually add the services to the builder.

Reply:
Huh?

Reply:
Oh dear.